src/main/scala/io/opentargets/etl/backend/target/Ensembl.scala

```diff
@@ -59,101 +58,93 @@ object Ensembl extends LazyLogging {
   }
 
   /** Returns dataframe with only one non-encoding gene per approvedSymbol. The other gene ids pointing to the same
-    * approvedSymbol are listed in `alternativeGenes`
+    * approvedSymbol are listed in `alternativeGenes`.
+    *
+    * The exception is that when more than one gene with the same gene ID points to different chromosomes, each are
+    * retained: eg.U2, U4, Y_RNA each have multiple EnsemblIds on different chromosomes.
     *
     * In cases where there is a gene on the canonical chromosome with the same approvedSymbol as the non-encoding gene,
     * all non-encoding genes will be listed as alternative genes to the gene on the canonical chromosome.
     *
-    * In cases where there are multiple gene ids, the longest will be chosen, with the longest being calculated as
-    * gene_end - gene_start. If there are multiple gene ids with the same length one will be chosen at random.
+    * In cases where there are multiple gene ids on non-canonical chromosomes, the longest will be chosen, with the
+    * longest being calculated as gene_end - gene_start. If there are multiple gene ids with the same length one will
+    * be chosen at random.
     *
     * All alternative gene ids which are reviewed are not included in the index will be included in the alternative id
     * field.
     */
   def selectBestNonReferenceGene(dataFrame: DataFrame): DataFrame = {
+
```

Comment:
@JarrodBaker this filtering looks a bit convoluted. Perhaps you might want to consider the use of windows for this. also, you can use the `abs()` col function too instead of multiplying by `-1` then you can use the length as an order desc in the window function so you don't artificially need to collect set and also we can skip the join at the bottom. Just composed boolean flag filtering

Reply:
I'm happy to work through a revision together @mkarmona . This refactor is replacing a what is probably a more elegant solution using windowing (but which was much slower and timed out the Travis tests). The convolution to my mind is being introduced because of the logic rules we are being asked to implement aren't necessarily logical, their biology. 

Multiplying by -1 was so I could use `array_sort` and get the order I wanted. I've just checked the Spark 3.0 docs and I see I can use a comparator (2.4 could only sort using natural order) so I'll fix that up.

Reply:
The best of biology at works